### ISSUE TITLE
Animation when opening keyboard is unsynchronized

### ISSUE THREAD

**Issue by magdabudzinskagoodsoft** [flutter/flutter#105687]
## Details

https://user-images.githubusercontent.com/87308455/172815472-2fa66bce-7f3f-4130-9bb0-a911cb51b125.mov

There is delay between keyboard appearing in iOS and beginning of Scaffold animation. Latest versions of Flutter improve this animation. It looks better now but still not perfectly. For iOS users it's very important to keep it much more native.

**Target Platform:** iOS
**Devices:** Physical devices

## Logs

flutter doctor -v 

```
Flutter (Channel stable, 3.0.1, on macOS 12.3.1 21E258 darwin-arm, locale en-PL)
 • Flutter version 3.0.1 
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fb57da5f94 (3 weeks ago), 2022-05-19 15:50:29 -0700
    • Engine revision caaafc5604
    • Dart version 2.17.1
    • DevTools version 2.12.2
```

**Comment by exaby73** [comment]
Hi @magdabudzinskagoodsoft, please provide a [minimal, reproducible example][1] and the output of `flutter doctor -v`.

[1]: https://stackoverflow.com/help/minimal-reproducible-example

**Comment by magdabudzinskagoodsoft** [comment]
This is code from the video above
```
class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        alignment: Alignment.bottomCenter,
        children: [
          ListView.builder(
            itemBuilder: ((context, index) {
              return ListTile(
                title: Text('Item $index'),
              );
            }),
            itemCount: 30,
          ),
          Container(height: 80, color: Colors.black12, child: const TextField())
        ],
      ),
    );
  }
}
```

**Comment by exaby73** [comment]
I can replicate this on stable and master.

<details><summary>Video</summary>

https://user-images.githubusercontent.com/50555895/172871784-700cf1c5-25a2-405a-8abd-2606207c6594.mov

</details>

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        alignment: Alignment.bottomCenter,
        children: [
          ListView.builder(
            itemBuilder: ((context, index) {
              return ListTile(
                title: Text('Item $index'),
              );
            }),
            itemCount: 30,
          ),
          Container(height: 80, color: Colors.black12, child: const TextField())
        ],
      ),
    );
  }
}
```

</details>

<details><summary><code>flutter doctor -v</code> (Stable)</summary>

```
[✓] Flutter (Channel stable, 3.0.1, on macOS 12.4 21F79 darwin-arm, locale en-US)
    • Flutter version 3.0.1 at /Users/nabeelparkar/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fb57da5f94 (3 weeks ago), 2022-05-19 15:50:29 -0700
    • Engine revision caaafc5604
    • Dart version 2.17.1
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/nabeelparkar/Library/Android/sdk/
    • Platform android-32, build-tools 32.1.0-rc1
    • ANDROID_SDK_ROOT = /Users/nabeelparkar/Library/Android/sdk/
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Ultimate Edition (version 2022.1.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 68.1.4
    • Dart plugin version 221.5787.37

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.42.0

[✓] Connected device (4 available)
    • V2050 (mobile)              • 3085425058000NM • android-arm64  • Android 12 (API 31)
    • sdk gphone64 arm64 (mobile) • emulator-5554   • android-arm64  • Android 12 (API 32) (emulator)
    • macOS (desktop)             • macos           • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)                • chrome          • web-javascript • Brave Browser 102.1.39.120

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

<details><summary><code>flutter doctor -v</code> (Master)</summary>

```
[✓] Flutter (Channel master, 3.1.0-0.0.pre.1189, on macOS 12.4 21F79 darwin-arm, locale en-US)
    • Flutter version 3.1.0-0.0.pre.1189 at /Users/nabeelparkar/fvm/versions/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f4604fe02b (65 minutes ago), 2022-06-09 11:13:10 +0300
    • Engine revision 3498d19583
    • Dart version 2.18.0 (build 2.18.0-170.0.dev)
    • DevTools version 2.14.0

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/nabeelparkar/Library/Android/sdk/
    • Platform android-32, build-tools 32.1.0-rc1
    • ANDROID_SDK_ROOT = /Users/nabeelparkar/Library/Android/sdk/
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Ultimate Edition (version 2022.1.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 68.1.4
    • Dart plugin version 221.5787.37

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.42.0

[✓] Connected device (4 available)
    • V2050 (mobile)              • 3085425058000NM • android-arm64  • Android 12 (API 31)
    • sdk gphone64 arm64 (mobile) • emulator-5554   • android-arm64  • Android 12 (API 32) (emulator)
    • macOS (desktop)             • macos           • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)                • chrome          • web-javascript • Brave Browser 102.1.39.120

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by justinmc** [comment]
This is the most recent PR that improved this: https://github.com/flutter/flutter/pull/99546.  CC @luckysmg.

Below is a video of this bug on master running in release mode on a physical device (iPhone 7), and it looks like it is indeed not perfectly in sync.  Especially as the keyboard closes, you can see a full white gap between the translucent bottom text fields and the keyboard at the bottom.

@luckysmg The video in your PR looks correct, do you know why it doesn't look as good in this case?


https://user-images.githubusercontent.com/389558/174138365-42b2e466-7b98-42e3-bf10-d0b81147624b.mp4

**Comment by luckysmg** [comment]
Hi @justinmc , I think this releated PR is https://github.com/flutter/engine/pull/29281 instead of that PR. 
I think in the engine side,we use `CADisplaylink` to track the animation interpolation, it is not a iOS native view that use the way like:`UIView.animate(block:)`,so it will have some diffrence...I think the animation curve is also right, see https://github.com/flutter/engine/pull/29281#issuecomment-962593711

**Comment by RichardFevrier** [comment]
Hi @luckysmg, does that mean we will probably never have a perfect synchronization with the keyboard on iOS?

**Comment by luckysmg** [comment]
Yes...I think so @RichardFevrier

**Comment by RichardFevrier** [comment]
Thanks for the answer @luckysmg, I'm sorry to hear that since text inputs are everywhere nowadays and since iDevices are not cheap their users are expecting a perfect UI. A big stop for flutter IMHO.

**Comment by RichardFevrier** [comment]
@luckysmg on the `main` branch in the `FlutterViewController` `-startKeyBoardAnimation`  `-animateWithDuration` I dont see any reference about the keyboard curve. Is it possible that this is the problem ? I made an Objc project on my side and my animation is in sync with the keyboard (i checked by slowing animation on the iOS simulator)

**Comment by luckysmg** [comment]
@RichardFevrier https://github.com/flutter/engine/pull/29281#issuecomment-962593711

**Comment by RichardFevrier** [comment]
@luckysmg thanks for pointing me to this comment, but I dont agree with your conclusion when you said its okay to not pass the curve (we can clearly see a blank during the animation on you second video, like the video made by @magdabudzinskagoodsoft on this issue). Here is what I got when I pass the curve:

https://user-images.githubusercontent.com/5154754/178105569-1dd982fe-aeca-461d-8e52-40368c836923.mp4

You can download my video and slow the playback in your prefered player, you will not see a blank between the keyboard and the red view.

**Comment by luckysmg** [comment]
Oh, That's nice. I will check it to comfirm..

And, Could you give me your flutter demo code😄? To comfirm the dart code is same with you

**Comment by RichardFevrier** [comment]
Here is my Objc demo code, I let you apply it on the `FlutterViewController.mm` 🙂 (that should be straightforward since you already have a UIView animateWithDuration: line 1319 on `main`)

[SyncKeyboardAnimation.zip](https://github.com/flutter/flutter/files/9076524/SyncKeyboardAnimation.zip)

**Comment by luckysmg** [comment]
Thx!😄

**Comment by RichardFevrier** [comment]
I didn't checked but do you think it will fix it for every components linked to the keyboard, like `Scaffold>FloatingActionButton`, etc... ? 🙂

**Comment by luckysmg** [comment]
Give a 5min..I will apply the curve and give you a preview video,Hold on😄

**Comment by luckysmg** [comment]
Emm....I m working on my flutter env..When I m done,I will notice you here😄...

**Comment by luckysmg** [comment]
Emm , I use your code and raise:
```
../../flutter/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1248:38: error: assigning to 'UIViewAnimationCurve' from incompatible type 'NSInteger' (aka 'long')
    keyboardTransitionAnimationCurve |= (UIViewAnimationCurve)keyboardTransitionAnimationCurve<<16;
                                     ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

And I modify your demo code.
I use code like below(I remove curve):
```objective-c
[UIView animateWithDuration:duration
                 animations:^{
                    CGRect frame = self.redSquare.frame;
                    frame.origin.y = self.view.frame.size.height - (notification.name == UIKeyboardWillHideNotification ? 0 : keyboardFrame.size.height) - 100;
                    self.redSquare.frame = frame;
               } completion:NULL];
```
The video(The same like you):

https://user-images.githubusercontent.com/49340347/178107791-230d95fb-68dd-47da-8067-537cec822c90.MP4

**Comment by luckysmg** [comment]
I have used the `[UIView setAnimationCurve]` to set the curve. But it is still have the white space....Not perfectly sync the ios keyboard...

**Comment by RichardFevrier** [comment]
Can you point me to a quick start guide to test a forked `flutter engine` please ? 🙂

**Comment by RichardFevrier** [comment]
@luckysmg do you recompile the engine everytime you update a single file in darwin platform to test it? I guess yes but I wish a miracle because it takes 2h+ for my laptop to compile the engine 😕

**Comment by luckysmg** [comment]
Of cource，I recompile every time



发自我的iPhone


------------------ Original ------------------
From: Richard Février ***@***.***&gt;
Date: Mon,Jul 11,2022 6:02 PM
To: flutter/flutter ***@***.***&gt;
Cc: WenJingRui ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [flutter/flutter] Animation when opening keyboard isunsynchronized (Issue #105687)





 
@luckysmg do you recompile the engine everytime you update a single file in darwin platform to test it? I guess yes but I wish a miracle because it takes 2h+ for my laptop to compile the engine 😕

**Comment by RichardFevrier** [comment]
Thanks 🙂 so I guess I will put that in pause for now, I can't take this much time on flutter.

**Comment by justinmc** [comment]
It's much faster after the first compilation if you're just changing one file.  But yeah it takes forever the first time :zzz:

**Comment by cutzmf** [comment]
@justinmc @RichardFevrier Guys, I appreciate your work!
So is there the chance to get it synchronized in later Flutter release?

**Comment by CoolDude53** [comment]
@luckysmg been following you on this and related issues / PRs.  Appreciate your work on it.

Saw https://github.com/flutter/engine/pull/29281 was merged, so I tried on master w/ physical iPhone 12 Pro on release mode. Still noticeably out of sync when keyboard is opening (less so on close). 

Any idea on getting parity with native? Happy to help on this if need be.

**Comment by luckysmg** [comment]
Hi @CoolDude53. You can see https://developer.apple.com/forums/thread/712762?login=true.

The view.layer.presentation() 's value is not accurate because there is no animation callback in iOS, so it is not a animation system....

**Comment by luckysmg** [comment]
Yes…..I think so.



发自我的iPhone


------------------ Original ------------------
From: Richard Février ***@***.***&gt;
Date: Sat,Jul 9,2022 6:58 AM
To: flutter/flutter ***@***.***&gt;
Cc: WenJingRui ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [flutter/flutter] Animation when opening keyboard isunsynchronized (Issue #105687)





 
Hi @luckysmg, does that mean we will probably never have a perfect synchronization with the keyboard on iOS?

**Comment by delfme** [comment]
@luckysmg thanks for pointing me here. Is there a workaround for the issue? Not having a perfect synchronization with keyboard makes apps feel cheap and buggy on ios. Users are so demanding that cannot live with this.

**Comment by delfme** [comment]
Adding here what people are expected to see. This is so basic and widely experienced that the unsynchronized animation is noticed in a snap and cannot be accepted.

https://user-images.githubusercontent.com/53510751/201531107-c4fecbe5-bf98-4aa9-9ae7-f613555825bb.MOV

**Comment by delfme** [comment]
@luckysmg maybe a workaround is possible. From all my tests it seems that textfield is always trying to catch up with the keyboard. Why we dont speed up textfield animation by a value which is % of keyboard speed? I guess if we nail this value by eye, then it should remain consistent ;)

**Comment by luckysmg** [comment]
The key point is that in iOS native side, I didn't find a API for catching interpolation during every animation frame.........
See https://developer.apple.com/forums/thread/712762?login=true

**Comment by delfme** [comment]
I got your point. Im suggesting that we might not need it if we reduce animation duration.

This should be Apple curve:
let ani = CASpringAnimation(keyPath: someKey)
ani.damping = 500
ani.stiffness = 1000
ani.mass = 3

Whatever the mass/stiffness/damping/velocity values, we cant make a good sync because, from iOS 15 on, Apple is opening keyword faster. This is why ui movement is always late compared to keyboard.

Fix should be to reduce current animation by 500ms or similar value. As soon as we make UI starts its movement with no delay compared to keyboard, we will have an almost perfect sync.

**Comment by luckysmg** [comment]
Emm I didn't get you `reduce current animation by 500ms` The animation is 0.25s(250ms) in total.

**Comment by luckysmg** [comment]
How should we get a appropriate value to reduce? Or Would you mind file a patch in `https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm`
And we can discuss about it ^_^

**Comment by luckysmg** [comment]
Later I will have some debuging to change the animation duration. And we can see the effect^_^

**Comment by delfme** [comment]
Thanks so much! By eye the delay looked about 500ms, that’s why I suggested that value. Can you try to set duration to 100ms or 50ms to see the difference? If “duration” is working, reducing it will make the fix. I say “if” because current animation takes more than declared 250ms and I fear that this value is not effective.

Would you be so kind as to also change a bit the demo? Most of tests are done with a lightweight UI. Unfortunately, flutter suffers of janks as soon as things become more realistic. For example, can we use a list of 200 items where each item comprises a fake image and a string? 🙏😊

**Comment by delfme** [comment]
> How should we get a appropriate value to reduce? Or Would you mind file a patch in `https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm`
> And we can discuss about it ^_^


Will have a look and revert.
In any event, the rationale is: whatever the curve, we want UI to start its movement almost exactly when keyboard starts its movement. To do that we can play with UI animation duration, the shorter its value the sooner UI will move.

**Comment by luckysmg** [comment]
I have tried to reduce the animation duration. But no use. Because it is already in an animation block. If I override its duration and curve. It show even more stranger...

**Comment by CoolDude53** [comment]
@luckysmg @delfme I've been debugging this when I can. I tried everything imaginable. What I got the best results with was using calculated animation points based on this project https://github.com/CosynPa/RevealSpringAnimation.

My implementation is similar to the existing; however, instead of using the presentationLayer's frame which causes the "catch up" affect (per [the apple thread](https://developer.apple.com/forums/thread/712762?login=true)), it uses a calculated animation completion percentage to more closely approximate where the keyboard actually is during the animation.

I'll try to start drafting a PR later today and will tag you to get your thoughts @luckysmg. It's not perfect (I don't think we can ever get there unless Apple opens up their keyboard animation curve or we rewrite the FVC), but it's a massive improvement.

Current Engine:
https://user-images.githubusercontent.com/5343432/201750194-1727b071-93de-442c-b4c6-dd994dd54ffc.MP4

My Engine:
https://user-images.githubusercontent.com/5343432/201750230-6702f947-4c1f-47fe-bbe2-2077ab5a9825.mov
